### PR TITLE
Fix passing bidirectional mask for multimodal

### PR DESCRIPTION
**Pull request by aireenmei**
# Description
The argument bidirectional_mask, which indicates the image placeholder tokens in the text input, was not passed correctly to text-image attention in some cases. 
After this change:

# Tests
* Verified by inserting jax.debug.print in the attention module.
* bidirectional_mask is passed correctly (tested) in: Gemma3 unscan, Llama4 unscan, Llama4 scan. 
* gemma3 scan is not tested, will test after gemma3 SFT is enabled.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/decoders.py [hunk 1]
```diff
@@ -686,21 +681,32 @@ def __call__(
...
-          scan_length = int(cfg.num_decoder_layers / cfg.inhomogeneous_layer_cycle_interval)
-          y, _ = self.scan_decoder_layers(cfg, RemattedBlockLayer, scan_length, "layers", mesh, **layer_kwargs)(
-              y,
-              decoder_segment_ids,
-              decoder_positions,
-              deterministic,
-              model_mode,
-          )
+            y, _ = self.scan_decoder_layers(
+                cfg, RemattedBlockLayer, scan_length, "layers", mesh, in_axes_len=5, **layer_kwargs
```
Comment (RissyRan) [inline MaxText/layers/decoders.py]: It seems this `in_axes_len=5` a hardcode value here? is that possible to pass from a config or based on existing configs?

Comment (aireenmei) [reply MaxText/layers/decoders.py]: Previously the in_axes is fixed to length of 4, corresponds to the #of arguments (https://screenshot.googleplex.com/9KEGayXhSNSWafE). I make it configurable with in_axes_len. In llama4 case, it's 5 as we pass 5 arguments.

Comment (gagika) [reply MaxText/layers/decoders.py]: to avoid magic numbers you could  change `scan_decoder_layers` to accept `in_axes` tuple

having broadcast_args 5 element or 4 element tuple for Llama4 or Gemma3 and 

in_axes_tuple = (nn.broadcast,) * len(broadcast_args)

 y, _ = self.scan_decoder_layers(
        cfg, RemattedBlockLayer, scan_length, "layers", mesh, in_axes=in_axes_tuple, **layer_kwargs
    )(y, *broadcast_args)

File: MaxText/multimodal_utils.py
Comment (hengtaoguo) [inline MaxText/multimodal_utils.py:1]: Thanks! I noticed there are two lines where we should update as a follow up to the previous "jax->numpy" migration in SFT. I've already been using these changes in my local branches. Could you help squeeze into your PR?

1. [Line here](https://github.com/AI-Hypercomputer/maxtext/blob/33339821e96327d25613ace9d28c7a8789698586/MaxText/multimodal_utils.py#L110) should be
`return np.array(image)`

2. [Line here](https://github.com/AI-Hypercomputer/maxtext/blob/33339821e96327d25613ace9d28c7a8789698586/MaxText/multimodal_utils.py#L325) should be
`image = np.array(resized_pil_image, dtype=np.float32)`

After the above change, I've run the below command on a v4-8 VM and it works fine:
```
python -m MaxText.decode MaxText/configs/base.yml model_name=gemma3-4b tokenizer_path=assets/tokenizer.gemma3 load_parameters_path=gs://hengtaoguo-maxtext-logs/checkpoints/gemma3-4b/vit/unscanned/2025-05-21-23-23-59/checkpoints/0/items per_device_batch_size=1 run_name=ht_test max_prefill_predict_length=272 max_target_length=300 steps=1 async_checkpointing=false scan_layers=false use_multimodal=true prompt=\'Describe\ image\ \<start_of_image\>\' image_path=\'MaxText/test_assets/test_image.jpg\' attention=\'dot_product\'
```

**Review by hengtaoguo — APPROVED** [review]
Thanks for catching this bug!

I rebased to your branch, and did a test run with [log](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-128-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-07-21-23-39-38-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev) using this [command](https://paste.googleplex.com/6258772888256512). Could you help apply the aforementioned fix and try a Llama4 decode? Let me know if it works on your side, thanks!

File: MaxText/multimodal_utils.py
Comment (aireenmei) [reply MaxText/multimodal_utils.py:1]: Added. Thanks Hengtao!

File: MaxText/layers/decoders.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/layers/decoders.py]: Thanks for the suggestion! I revised and add refactored the code.

**Comment by aireenmei** [comment]
> Thanks for catching this bug!
> 
> I rebased to your branch, and did a test run with [log](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-128-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-07-21-23-39-38-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev) using this [command](https://paste.googleplex.com/6258772888256512). Could you help apply the fix I mentioned, and try a Llama4 decode? Let me know if it works on your side, thanks!

I checked the revised and rebased branch for decode. The result is the same ([log](https://cloudlogging.app.goo.gl/x1SuvyXycR8uo6b39))

File: MaxText/multimodal_utils.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def load_image_from_path(image_path):
   try:
     image = Image.open(image_path).convert("RGB")
     image.load()  # Load image data to catch errors early
-    return jnp.asarray(np.array(image))
+    return np.asarray(np.array(image))
```
Comment (gagika) [inline MaxText/multimodal_utils.py]: will this work instead?
`np.asarray(image)` or `np.array(image)` ?

Comment (aireenmei) [reply MaxText/multimodal_utils.py]: Change it to `np.asarray(image)`. Thanks for catching!

File: MaxText/layers/decoders.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/layers/decoders.py]: One note is that pipline_stage_module is initialized in Decoder setup(), it needs the length of broadcast_args, which is defined at `__call()__`. Previously it was hardcoded. I updated the PR using Gemini's suggestion of using python inspect module to know the args in `__call()__`. This results in somewhat "hardcoding" the args orders in `Pipeline.__call()__`, I added comments. What do you think? @gobbleturk

File: MaxText/layers/decoders.py [hunk 3]
```diff
@@ -458,8 +454,24 @@ def get_layer_to_pipeline(blocks, cfg):
     if cfg.num_layers_per_pipeline_stage == 1:
       stage_module = base_stage(config=cfg, mesh=self.mesh, quant=self.quant)
     elif cfg.scan_layers_per_stage:
+      # Dynamically determine the number of broadcast arguments.
+      # Assuming the broadcasted args are between 'inputs' and 'partition_spec' when passed to pipeline.Pipeline.__call__.
+      sig = inspect.signature(pipeline.Pipeline.__call__)
```
Comment (gobbleturk) [inline MaxText/layers/decoders.py]: oh my lawd

Comment (aireenmei) [reply MaxText/layers/decoders.py]: Looks like the code complexity is not great. I reverse it back to hard coded in_axes_len for get_pipeline_stage_module

File: MaxText/layers/decoders.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/layers/decoders.py]: After talking offline with @gobbleturk and @RissyRan, we don't like the complexity of using the inspect module. I reverse to hardcoded in_axes_len for get_pipeline_stage_module
